### PR TITLE
Moving webhook to its own service

### PR DESCRIPTION
**Pull request by AndreaFrancis**
Fix for https://github.com/huggingface/dataset-viewer/issues/2848

### CODE REVIEW

**Comment by AndreaFrancis** [comment]
e2e is still running but I would love an early review since it is a critical issue

File: services/webhook/tests/__init__.py [hunk 1]
```diff
@@ -0,0 +1,2 @@
+# SPDX-License-Identifier: Apache-2.0
+# Copyright 2022 The HuggingFace Authors.
```
Comment (severo) [inline services/webhook/tests/__init__.py:2]: nit: 2024

File: services/webhook/tests/routes/__init__.py
Comment (severo) [inline services/webhook/tests/routes/__init__.py:1]: nit: missing header

**Comment by AndreaFrancis** [comment]
> e2e is still running but I would love an early review since it is a critical issue

Looks like e2e passed

File: chart/templates/services/webhook/ingress.yaml [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        paths:
+          - backend:
+              service:
+                name: "{{ include "name" . }}-webhook"
+                port:
+                  name: http
+            path: /
+            pathType: Prefix
+          {{ if hasKey $annotations "alb.ingress.kubernetes.io/actions.openapi-redirect" -}}
+          - path: /openwebhook.json
```
Comment (severo) [inline chart/templates/services/webhook/ingress.yaml:23]: this part should be deleted (it's only to serve openapi.json)

File: chart/env/prod.yaml [hunk 3]
```diff
@@ -537,3 +541,31 @@ workers:
...
+  # (2 x $num_cores) + 1
+  # https://docs.gunicorn.org/en/stable/design.html#how-many-workers
+  uvicornNumWorkers: "9"
+  nodeSelector:
+    role-datasets-server-webhook: "true"
+  tolerations:
+    - key: "huggingface.co/datasets-server-webhook"
+      operator: "Exists"
+      effect: "NoSchedule"
+  replicas: 20
```
Comment (severo) [inline chart/env/prod.yaml:556]: I don't think we need so many of them

File: chart/env/prod.yaml [hunk 4]
```diff
@@ -537,3 +541,31 @@ workers:
...
+  service:
+    type: NodePort
+  ingress:
+    enabled: true
+    annotations:
+      alb.ingress.kubernetes.io/target-node-labels: role-datasets-server-webhook=true
+      alb.ingress.kubernetes.io/actions.openapi-redirect: '{"Type":"redirect","RedirectConfig":{"Host":"raw.githubusercontent.com","Path":"/huggingface/dataset-viewer/main/docs/source/openapi.json","Port":"443","Protocol":"HTTPS","Query":"#{query}","StatusCode":"HTTP_302"}}'
```
Comment (severo) [inline chart/env/prod.yaml:563]: not needed

File: chart/env/prod.yaml [hunk 5]
```diff
@@ -537,3 +541,31 @@ workers:
...
+      alb.ingress.kubernetes.io/actions.metrics-unauthorized: '{"type":"fixed-response","fixedResponseConfig":{"contentType":"text/plain","statusCode":"401","messageBody":"401 Unauthorized"}}'
+  resources:
+    requests:
+      cpu: 4
+      memory: "14Gi"
```
Comment (severo) [inline chart/env/prod.yaml:568]: I don't think we need that quantity of RAM (hopefully)

File: chart/env/prod.yaml [hunk 6]
```diff
@@ -537,3 +541,31 @@ workers:
...
+    limits:
+      cpu: 4
+      memory: "14Gi"
```
Comment (severo) [inline chart/env/prod.yaml:571]: same

File: chart/env/staging.yaml [hunk 7]
```diff
@@ -320,3 +325,21 @@ workers:
...
+
+webhook:
+  uvicornNumWorkers: "1"
+  replicas: 1
+  service:
+    type: NodePort
+  ingress:
+    enabled: true
+    annotations:
+      alb.ingress.kubernetes.io/actions.openapi-redirect: '{"Type":"redirect","RedirectConfig":{"Host":"raw.githubusercontent.com","Path":"/huggingface/dataset-viewer/main/docs/source/openapi.json","Port":"443","Protocol":"HTTPS","Query":"#{query}","StatusCode":"HTTP_302"}}'
```
Comment (severo) [inline chart/env/staging.yaml:337]: remove

File: services/webhook/Makefile [hunk 8]
```diff
@@ -0,0 +1,32 @@
+# environment variables for the commands (docker compose, poetry)
+export COMPOSE_PROJECT_NAME := webhook
+export MONGO_PORT := 27031
```
Comment (severo) [inline services/webhook/Makefile:3]:
```suggestion
export MONGO_PORT := 27036
```

File: services/webhook/pyproject.toml [hunk 9]
```diff
@@ -0,0 +1,66 @@
+[tool.poetry]
+authors = ["Sylvain Lesage <sylvain.lesage@huggingface.co>"]
+description = "REST API app"
```
Comment (severo) [inline services/webhook/pyproject.toml:3]:
```suggestion
description = "webhook app"
```

File: services/webhook/pyproject.toml [hunk 10]
```diff
@@ -0,0 +1,66 @@
+[tool.poetry]
+authors = ["Sylvain Lesage <sylvain.lesage@huggingface.co>"]
```
Comment (severo) [inline services/webhook/pyproject.toml:2]: you can change this

File: services/webhook/pyproject.toml [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+name = "webhook"
+version = "0.1.3"
+license = "Apache-2.0"
+
+[tool.poetry.dependencies]
+python = "3.9.18"
+environs = "^9.5.0"
+jsonschema = "^4.17.0"
+libapi = {path = "../../libs/libapi", develop = true}
+soundfile = ">=0.12.1"
```
Comment (severo) [inline services/webhook/pyproject.toml:13]: not sure we need this (same in the docker file) but...

File: services/webhook/src/webhook/__init__.py [hunk 12]
```diff
@@ -0,0 +1,2 @@
+# SPDX-License-Identifier: Apache-2.0
+# Copyright 2022 The HuggingFace Authors.
```
Comment (severo) [inline services/webhook/src/webhook/__init__.py:2]: 2024

File: chart/templates/services/webhook/ingress.yaml [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+            pathType: Exact
+            backend:
+              service:
+                name: openapi-redirect
+                port:
+                  name: use-annotation
+          {{- end }}
+          {{ if hasKey $annotations "alb.ingress.kubernetes.io/actions.metrics-unauthorized" -}}
+          - path: /metrics
```
Comment (severo) [inline chart/templates/services/webhook/ingress.yaml:32]: we have to delete this too (see /rows)

File: chart/templates/services/webhook/ingress.yaml [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+  rules:
+    - host: {{ include "datasetsServer.ingress.hostname" . }}
+      http:
+        paths:
+          - backend:
+              service:
+                name: "{{ include "name" . }}-webhook"
+                port:
+                  name: http
+            path: /
```
Comment (severo) [inline chart/templates/services/webhook/ingress.yaml:20]:
```suggestion
            path: /webhook
```

File: chart/env/prod.yaml [hunk 15]
```diff
@@ -537,3 +541,31 @@ workers:
...
+    - key: "huggingface.co/datasets-server-webhook"
+      operator: "Exists"
+      effect: "NoSchedule"
+  replicas: 20
+  service:
+    type: NodePort
+  ingress:
+    enabled: true
+    annotations:
+      alb.ingress.kubernetes.io/target-node-labels: role-datasets-server-webhook=true
```
Comment (severo) [inline chart/env/prod.yaml:562]: missing group.order: 

```suggestion
      alb.ingress.kubernetes.io/group.order: "5"
      alb.ingress.kubernetes.io/target-node-labels: role-datasets-server-webhook=true
```

File: chart/env/prod.yaml [hunk 16]
```diff
@@ -537,3 +541,31 @@ workers:
...
+      alb.ingress.kubernetes.io/actions.openapi-redirect: '{"Type":"redirect","RedirectConfig":{"Host":"raw.githubusercontent.com","Path":"/huggingface/dataset-viewer/main/docs/source/openapi.json","Port":"443","Protocol":"HTTPS","Query":"#{query}","StatusCode":"HTTP_302"}}'
+      alb.ingress.kubernetes.io/actions.metrics-unauthorized: '{"type":"fixed-response","fixedResponseConfig":{"contentType":"text/plain","statusCode":"401","messageBody":"401 Unauthorized"}}'
```
Comment (severo) [inline chart/env/prod.yaml:564]: we can remove too

File: chart/env/staging.yaml [hunk 17]
```diff
@@ -320,3 +325,21 @@ workers:
...
+      alb.ingress.kubernetes.io/actions.metrics-unauthorized: '{"type":"fixed-response","fixedResponseConfig":{"contentType":"text/plain","statusCode":"401","messageBody":"401 Unauthorized"}}'
```
Comment (severo) [inline chart/env/staging.yaml:338]: we can remove

File: chart/env/prod.yaml [hunk 18]
```diff
@@ -537,3 +541,31 @@ workers:
...
         cpu: 2
         memory: "1Gi"
+
+webhook:
+  # Number of uvicorn workers for running the application
+  # (2 x $num_cores) + 1
+  # https://docs.gunicorn.org/en/stable/design.html#how-many-workers
+  uvicornNumWorkers: "9"
+  nodeSelector:
+    role-datasets-server-webhook: "true"
```
[the remark this answers is not in the document]
Comment (rtrompier) [reply chart/env/prod.yaml:551]: LGTM

File: chart/env/prod.yaml (hunk 5 above)
Comment (AndreaFrancis) [reply chart/env/prod.yaml]: I put 4Gi, feel free to suggest another value if needed

File: services/webhook/pyproject.toml (hunk 11 above)
Comment (AndreaFrancis) [reply services/webhook/pyproject.toml]: I remove it

File: chart/env/prod.yaml (hunk 3 above)
Comment (AndreaFrancis) [reply chart/env/prod.yaml]: I put 4 replicas, but I'm not sure if that is enough, can you advise?

File: chart/env/staging.yaml [hunk 19]
```diff
@@ -320,3 +325,20 @@ workers:
...
+
+webhook:
+  uvicornNumWorkers: "1"
+  replicas: 1
+  service:
+    type: NodePort
+  ingress:
+    enabled: true
+    annotations:
+      alb.ingress.kubernetes.io/group.order: "4"
```
Comment (AndreaFrancis) [inline chart/env/staging.yaml:337]: I followed rows and search annotations, but I am not sure if it is ok

File: chart/env/prod.yaml (hunk 3 above)
Comment (severo) [reply chart/env/prod.yaml]: let's monitor

**Review by severo — APPROVED** [review]
Excellent. Let's go!

**Pull request merged**